R/teal_slice.R

```diff
@@ -0,0 +1,431 @@
+  if (is.numeric(i) & max(i) > length(x)) stop("subscript out of bounds")
+  if (is.character(i) & !all(is.element(i, names(x)))) stop("subscript out of bounds")
+
+  y <- NextMethod("[")
+  attrs <- attributes(x)
+  attrs$names <- attrs$names[i]
+  attributes(y) <- attrs
+  excludes <- unique(unlist(vapply(y, function(ts) ts[["dataname"]], character(1L))))
+  attr(y, "exclude") <- Filter(Negate(is.null), attr(x, "exclude")[excludes])
+  y
+}
+
+
+# concatenate method for teal_slices
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+c.teal_slices <- function(...) {
+  x <- list(...)
+  excludes <- lapply(x, attr, "exclude")
+  names(excludes) <- NULL
+  excludes <- unlist(excludes, recursive = FALSE)
+  excludes <- excludes[!duplicated(names(excludes))]
+  count_types <- lapply(x, attr, "count_type")
+  count_types <- unique(unlist(count_types))
+  checkmate::assert_string(count_types)
+
+  do.call(filter_settings, c(unlist(x, recursive = FALSE), list(exclude = excludes, count_type = count_types)))
+}
+
+
+# print method for teal_slices
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+print.teal_slices <- function(x, ...) {
+  f <- attr(x, "exclude")
+  ct <- attr(x, "count_type")
+  for (i in seq_along(x)) {
+    ind <- names(x)[i]
+    if (is.null(ind)) ind <- sprintf("[[%d]]", i)
+    cat(ind, "\n")
+    print(x[[i]], ...)
+  }
+  cat("\nnon-filterable variables:")
+  if (is.list(f) & length(f) == 0L) {
+    cat(" none\n")
+  } else {
+    cat("\n")
+    for (i in seq_along(f)) {
+      cat(sprintf(" $ %s: %s", names(f)[i], toString(f[[i]])), "\n")
+    }
+  }
+  cat(sprintf("\ncount type: %s", ct), "\n")
+}
+
+
+# get slices where feature is value
+#' @rdname teal_slice
+#' @keywords internal
+#'
+extract_by_feat <- function(tss, feature, value) {
+  checkmate::assert_class(tss, "teal_slices")
+  Filter(function(x) x[[feature]] == value, tss)
+}
+
+
+# get feature from all slices
+#' @rdname teal_slice
+#' @keywords internal
+#'
+extract_feat <- function(tss, feature) {
+  checkmate::assert_class(tss, "teal_slices")
+  lapply(tss, function(x) x[[feature]])
+}
+
+
+# get slices where logical predicate is TRUE
+#' @rdname teal_slice
+#' @keywords internal
+#'
+extract_fun <- function(tss, expr) {
+  checkmate::assert_class(tss, "teal_slices")
+  expr <- substitute(expr)
+  checkmate::assert_class(expr, "call")
+  Filter(function(x) isTRUE(eval(expr, x)), tss)
+}
+
+
+# string version
+#' @rdname teal_slice
+#' @keywords internal
+#'
+extract_fun_s <- function(tss, expr) {
+  checkmate::assert_class(tss, "teal_slices")
+  checkmate::assert_string(expr)
+  expr <- str2lang(expr)
+  Filter(function(x) isTRUE(eval(expr, x)), tss)
+}
+
+
+# name teal_slices according to value of respective feature in slices
+# possibly useful in set_filter_state_impl but needs safeguards
+#' @rdname teal_slice
+#' @keywords internal
+#'
+name_slices <- function(tss, feature) {
```

Comment:
I don't see this used anywhere. If this is a have just in case, I'd encourage removing it. In my experience this just creates confusion down the road.

Reply:
Yup, some of these helpers will go away as soon as the core of the PR settles. I will go through the code and try to use as few as possible. Specifically, I don't think `extract_fun` is needed anymore as we don't have nested lists in a `teal_slice` so we don't have to use e.g. `extras$experiment == "exp_name"`.

Reply:
removed